### ISSUE TITLE
Better support long strings

### ISSUE THREAD

**Issue by Mardak** [mozilla/activity-stream#3244]
Here's a screenshot of the longest strings of any translated locale:
![screen shot 2017-08-25 at 11 16 43 pm](https://user-images.githubusercontent.com/438537/29739152-ef0f79a6-89eb-11e7-898f-5610ea843823.png)

Worst offenders look to be:
- manual migration description (here 3 lines but if the buttons were smaller, it would have been 2)
- preference option header (wraps to below checkbox)
- context menu options (long ones could potentially clip off edge of display)
- preference header (wraps but looks like it might be expecting 1 line?)
- topics list (at the very bottom seems like there's potential of overlap depending on topics)

@flodolo anything else to watch out for and fix?

Here's the diff I used to pick out the strings:
```diff
diff --git a/system-addon/lib/LocalizationFeed.jsm b/system-addon/lib/LocalizationFeed.jsm
--- a/system-addon/lib/LocalizationFeed.jsm
+++ b/system-addon/lib/LocalizationFeed.jsm
@@ -40,6 +40,17 @@ this.LocalizationFeed = class LocalizationFeed {
     // Start with default (first) locale then merge in strings of better locales
     const strings = Object.assign({}, ...locales.map(l => this.allStrings[l]));
 
+    // Replace strings with the longest translated string
+    for (const key of Object.keys(strings)) {
+      for (const locale of Object.keys(this.allStrings)) {
+        const other = this.allStrings[locale][key];
+        if (other && other.length > strings[key].length) {
+          // Identify the locale by replacing the first few letters
+          strings[key] = `${locale}:${other.slice(locale.length + 1)}`;
+        }
+      }
+    }
+
     // Use the best (last) locale as the primary locale
     const locale = locales.pop();
     this.store.dispatch(ac.BroadcastToContent({
```

**Comment by Mardak** [comment]
I suppose the 3-line manual migration isn't super critical (?) as here's what dsb looks like with its own strings for buttons:
<img width="755" alt="image" src="https://user-images.githubusercontent.com/438537/29739252-00508f1e-89ee-11e7-997a-0c47e05862fe.png">

**Comment by flodolo** [comment]
Wow, thanks for doing this, it's a really smart approach. 

Looking at ne-NP marked as the longest translation for *Dismiss*, I wonder if the code is doing the right thing though (bg, for example, has `Отхвърляне`).
```
'खारेज गर्नुहोस्'.length = 15
'Отхвърляне'.length = 10
'💩'.length = 2
```

I'm also one of the offenders with the translation for **Pin** (I hate that verb…).

I don't think we should worry too much about the context menu breaking? It definitely looks unbalanced with very long and very short strings together, but that's common in Firefox too

![schermata 2017-08-26 alle 08 45 02](https://user-images.githubusercontent.com/3644868/29739313-e9f25b74-8a3a-11e7-8a91-3b73efc16154.png)

**Comment by Mardak** [comment]
Oh. Wow 💩💩💩… Well here's a really quick hack:
```js
diff --git a/system-addon/lib/LocalizationFeed.jsm b/system-addon/lib/LocalizationFeed.jsm
--- a/system-addon/lib/LocalizationFeed.jsm
+++ b/system-addon/lib/LocalizationFeed.jsm
@@ -47 +47,2 @@ this.LocalizationFeed = class LocalizationFeed {
-        if (other && other.length > strings[key].length) {
+        const check = other && other.replace(/{\w+}/g, "");
+        if (other && check.toLowerCase() != check.toUpperCase() && other.length > strings[key].length) {
```
![screen shot 2017-08-26 at 9 23 40 am](https://user-images.githubusercontent.com/438537/29743148-a7d1f14a-8a40-11e7-8e9e-1538e6c0c159.png)

**Comment by flodolo** [comment]
> I suppose the 3-line manual migration isn't super critical (?) as here's what dsb looks like with its own strings for buttons:

I think it would be nice to make sure some vertical padding is available (looking at the first screenshot).

I've just realized you're doing text transformation on the section titles, make sure the language is correctly set.
https://bugzilla.mozilla.org/show_bug.cgi?id=1175132

I can't think of anything else. The only thing that look bad is the checkbox in preferences, with text wrapping below the checkbox itself.

**Comment by Mardak** [comment]
Ah yup, there's a lang tag on the `<html>` element 
<img width="490" alt="image" src="https://user-images.githubusercontent.com/438537/29747102-32c7bcdc-8aa3-11e7-9031-7c64e0110d64.png">

**Comment by Mardak** [comment]
Fixes for the original (wrong length calculation) strings with 3-line migration and wrapped pref:
![screen shot 2017-08-26 at 10 12 36 pm](https://user-images.githubusercontent.com/438537/29747337-5df1e2c6-8aac-11e7-8822-9eab39ae3b57.png)

**Comment by flodolo** [comment]
It definitely looks better. 

Worth trying to put a longer text in the suggested topics section to see what happens with the layout? ga-IE is at the limit, longer topics (or a larger number of topics) could exceed it.

**Comment by Mardak** [comment]
@csadilek do you know what's the expected behavior of long / too many topics to fit when other strings are long? Is it supposed to always be displayed on one line, so something should get cropped/ellipses?

If just the topics list left to fix, let's just do it as part of #3238 and close this one when #3246 lands.

**Comment by Mardak** [comment]
Here's how things currently look with long strings and many topics:
<img width="767" alt="image" src="https://user-images.githubusercontent.com/438537/29751442-41525dd4-8b02-11e7-9112-ab49476cac2b.png">

and RTL:
<img width="760" alt="image" src="https://user-images.githubusercontent.com/438537/29751437-2fe91132-8b02-11e7-8a7f-dda6e1f6bf6f.png">

And with fixes for #3238:
<img width="763" alt="image" src="https://user-images.githubusercontent.com/438537/29751452-653800be-8b02-11e7-9fc8-2294998c5d61.png">

RTL:
<img width="765" alt="image" src="https://user-images.githubusercontent.com/438537/29751457-70c95f9a-8b02-11e7-8779-6eb2e6a48c0c.png">
